x-pack/packages/ai-infra/inference-common/src/output/api.ts

```diff
@@ -6,39 +6,141 @@
+ *   `,
+ * });
+ *
+ * // output is properly typed from the provided schema
+ * const { animals } = response.output;
+ * ```
  */
 export type OutputAPI = <
   TId extends string = string,
-  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
 >(
-  id: TId,
-  options: {
-    connectorId: string;
-    system?: string;
-    input: string;
-    schema?: TOutputSchema;
-    previousMessages?: Message[];
-    functionCalling?: FunctionCallingMode;
-  }
-) => OutputResponse<TId, TOutputSchema>;
+  options: OutputOptions<TId, TOutputSchema, TStream>
+) => OutputCompositeResponse<TId, TOutputSchema, TStream>;
+
+/**
+ * Options for the {@link OutputAPI}
+ */
+export interface OutputOptions<
+  TId extends string = string,
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
+> {
+  /**
+   * The id of the operation.
+   */
+  id: TId;
+  /**
+   * The ID of the connector to use.
+   * Must be a genAI compatible connector, or an error will be thrown.
+   */
+  connectorId: string;
+  /**
+   * Optional system message for the LLM.
+   */
+  system?: string;
+  /**
+   * The prompt for the LLM.
+   */
+  input: string;
+  /**
+   * The {@link ToolSchema} the response from the LLM should adhere to.
+   */
+  schema?: TOutputSchema;
+  /**
+   * Previous messages in the conversation.
+   * If provided, will be passed to the LLM in addition to `input`.
+   */
+  previousMessages?: Message[];
+  /**
+   * Function calling mode, defaults to "native".
+   */
+  functionCalling?: FunctionCallingMode;
+  /**
+   * Set to true to enable streaming, which will change the API response type from
+   * a single promise to an event observable.
+   *
+   * Defaults to false.
+   */
+  stream?: TStream;
+}
+
+/**
+ * Composite response type from the {@link OutputAPI},
+ * which can be either an observable or a promise depending on
+ * whether API was called with stream mode enabled or not.
+ */
+export type OutputCompositeResponse<
+  TId extends string = string,
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
+> = TStream extends true
+  ? OutputStreamResponse<TId, TOutputSchema>
+  : Promise<
+      OutputResponse<
+        TId,
+        TOutputSchema extends ToolSchema ? FromToolSchema<TOutputSchema> : undefined
+      >
+    >;
+
+/**
+ * Response from the {@link OutputAPI} when streaming is not enabled.
+ */
+export interface OutputResponse<TId extends string = string, TOutput extends Output = Output> {
+  /**
+   * The id of the operation, as provided as input
+   */
+  id: TId;
+  /**
+   * The task output, following the schema specified as input
+   */
+  output: TOutput;
+  /**
+   * Potential text content provided by the LLM,
+   * if it was provided in addition to the tool call
+   */
+  content: string;
```

Comment:
question: is `content` always defined, even when text content is not provided by the LLM?

Reply:
There's 2 scenarios where `content` is defined:
- if the output API is called without providing a schema (in that case the LLM will always answer with text content)
- when a schema is provided, the LLM (depending on the provider/llm) can include text content in addition to the tool call

But to answer the question yes, atm `content` is always defined. It will be an empty string if the LLM did not return text content. This could potentially be revisited (e.g making it defined as optional on the response type and returning undefined instead of an empty string), but I'd rather keep that kind of changes for later (e.g. we may want to "specialize" this API more by enforcing the schema parameter, that kind of things, and I'd like to do all those things at the same time)

Reply:
That makes sense to me. I just wanted to make sure that the interface matched reality, and it seems that it does